src/DynamoCore/Graph/Nodes/NodeModel.cs

```diff
@@ -1083,6 +1084,103 @@ protected NodeModel(IEnumerable<PortModel> inPorts, IEnumerable<PortModel> outPo
             RaisesModificationEvents = true;
         }
 
+        /// <summary>
+        /// here we try to find the correct port names and tooltips.
+        /// ideally we'd use the runtime information to correctly update or localize
+        /// the port info, if we can't find it for this port we fallback to the deserialized data.
+        /// phases are:
+        /// 1. try to extract port data attributes from the nodeModel Type.
+        /// 2. create a temp node using the default constructor and extract port info.
+        /// 3. for variable input nodes - call the getTooltip method for each port. 
+        /// TODO this is an alternative, not step 4. for any node where portAttributeData count does not match port count, give up and use deserialized data.
+        /// eventually remove step 2.
+        /// </summary>
+        /// <param name="inPorts"></param>
+        /// <param name="outPorts"></param>
+        private void FindPortData(IEnumerable<PortModel> inPorts, IEnumerable<PortModel> outPorts)
+        {
+            //try to get portData from attributes on the nodeModel type.
+            //to assign the correct tooltips.
+            //This might fail because some nodes don't use attributes to define 
+            //port data and instead add them using PortModel constructors.
+            var inportDatas = GetPortDataFromAttributes(PortType.Input).ToList();
+            var outportDatas = GetPortDataFromAttributes(PortType.Output).ToList();
+
+            // if the above attempt failed, for example because the node type does not have any port data attributes
+            // then try other approaches
+            if (inportDatas.Count() != inPorts.Count() || outportDatas.Count() != outPorts.Count())
```

Comment:
so if port data (tooltip/description) is not supplied for InputPorts but is supplied for OutputPorts...then we try fallback for OutputPorts too ?
is it possible to do it only for the ports (input/output) that do not have the required data ?

Reply:
Same question here.